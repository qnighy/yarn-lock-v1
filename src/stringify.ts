/* eslint-disable @typescript-eslint/ban-types */
// Based on https://github.com/yarnpkg/yarn/blob/v1.22.10/src/lockfile/stringify.js

const INDENT = "  ";

export function stringify(obj: object): string {
  const lines: string[] = [];
  stringifyPush(obj, "", true, lines);
  return lines.join("");
}

function stringifyPush(
  obj: object,
  indent: string,
  topLevel: boolean,
  lines: string[]
) {
  const identicalKeys = new Map<object, string[]>();
  for (const [k, v] of Object.entries(obj)) {
    if (typeof v === "object" && v !== null) {
      const keys = identicalKeys.get(v);
      if (keys) keys.push(k);
      else identicalKeys.set(v, [k]);
    }
  }

  let isFirst = true;
  const addSeparator = () => {
    if (!isFirst && topLevel) {
      lines.push("\n");
    }
    isFirst = false;
  };
  for (const [primaryKey, v] of sortEntriesWithPriority<unknown>(
    Object.entries(obj)
  )) {
    switch (typeof v) {
      case "boolean":
      case "number":
      case "string":
        addSeparator();
        lines.push(`${indent}${wrapKey(primaryKey)} ${wrapKey(v)}\n`);
        continue;
      case "function":
      case "object":
        if (v === null) continue;
        break;
      case "undefined":
        continue;
      default:
        // symbol or bigint
        throw new Error(`Cannot stringify ${typeof v}`);
    }

    const keys = identicalKeys.get(v);
    if (keys === undefined) continue;
    identicalKeys.delete(v);

    const joinedKeys = keys.sort().map(wrapKey).join(", ");
    addSeparator();
    lines.push(`${indent}${joinedKeys}:\n`);
    stringifyPush(v, indent + INDENT, false, lines);
  }
}

function shouldWrapKey(s: string): boolean {
  return /^true|^false|^[0-9]|[:\s\n\\",[\]]/.test(s) || !/^[a-zA-Z]/.test(s);
}

export function wrapKey(s: string | boolean | number): string {
  if (typeof s === "boolean" || typeof s === "number" || shouldWrapKey(s)) {
    return JSON.stringify(s);
  } else {
    return s;
  }
}

const priorities: Readonly<Record<string, number>> = {
  name: 1,
  version: 2,
  uid: 3,
  resolved: 4,
  integrity: 5,
  registry: 6,
  dependencies: 7,
};

export function sortEntriesWithPriority<T>(keys: [string, T][]): [string, T][] {
  return keys.sort(([ak], [bk]) => cmpWithPriority(ak, bk));
}

export function sortWithPriority(keys: string[]): string[] {
  return keys.sort(cmpWithPriority);
}

function cmpWithPriority(a: string, b: string): number {
  const ap = Object.prototype.hasOwnProperty.call(priorities, a)
    ? priorities[a]
    : 100;
  const bp = Object.prototype.hasOwnProperty.call(priorities, b)
    ? priorities[b]
    : 100;
  if (ap !== bp) return ap - bp;
  return cmpAlpha(a, b);
}

// UCS-2 (UTF-16) lexcographical ordering
function cmpAlpha(a: string, b: string): number {
  const minLen = Math.min(a.length, b.length);
  for (let i = 0; i < minLen; ++i) {
    const ai = a.charCodeAt(i);
    const bi = b.charCodeAt(i);
    if (ai !== bi) return ai - bi;
  }
  return a.length - b.length;
}
