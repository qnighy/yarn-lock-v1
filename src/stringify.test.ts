import { describe, expect, it } from "@jest/globals";

import { sortWithPriority, stringify, wrapKey } from "./stringify";

describe("stringify", () => {
  it("stringifies an empty object", () => {
    const s = stringify({});
    expect(s).toMatchSnapshot();
  });

  it("stringifies an object with a single key", () => {
    const s = stringify({
      foo: {},
    });
    expect(s).toMatchSnapshot();
  });

  it("stringifies simple values", () => {
    const s = stringify({
      foo: {
        numeric: 42,
        snumeric: "42",
        boolean: false,
        sboolean: "true",
        s: "baz",
        s2: "baz baz",
      },
    });
    expect(s).toMatchSnapshot();
  });

  it("stringifies package info", () => {
    const s = stringify({
      "call-bind@^1.0.2": {
        version: "1.0.2",
        resolved: "https://registry.yarnpkg.com/call-bind/-/call-bind-1.0.2.tgz#b1d4e89e688119c3c9a903ad30abb2f6a919be3c",
        integrity: "sha512-7O+FbCihrB5WGbFYesctwmTKae6rOiIzmz1icreWJ+0aA7LJfuqhEso2T9ncpcFtzMQtzXf2QGGueWJGTYsqrA==",
        dependencies: {
          "function-bind": "^1.1.1",
          "get-intrinsic": "^1.0.2",
        },
      },
      "has-symbols@^1.0.1": {
        version: "1.0.2",
        resolved: "https://registry.yarnpkg.com/has-symbols/-/has-symbols-1.0.2.tgz#165d3070c00309752a1236a479331e3ac56f1423",
        integrity: "sha512-chXa79rL/UC2KlX17jo3vRGz0azaWEx5tGqZg5pO3NUyEJVB17dMruQlzCCOfUvElghKcm5194+BCRvi2Rv/Gw==",
      },
      "is-regex@^1.1.2": {
        version: "1.1.2",
        resolved: "https://registry.yarnpkg.com/is-regex/-/is-regex-1.1.2.tgz#81c8ebde4db142f2cf1c53fc86d6a45788266251",
        integrity: "sha512-axvdhb5pdhEVThqJzYXwMlVuZwC+FF2DpcOhTS+y/8jVq4trxyPgfcwIxIKiyeuLlSQYKkmUaPQJ8ZE4yNKXDg==",
        dependencies: {
          "call-bind": "^1.0.2",
          "has-symbols": "^1.0.1",
        },
      },
    });
    expect(s).toMatchSnapshot();
  });

  it("doesn't unify same primitives", () => {
    const s = stringify({
      foo: {
        foo: 42,
        bar: 42,
      }
    });
    expect(s).toMatchSnapshot();
  });

  it("doesn't unify equivalent objects", () => {
    const s = stringify({
      foo: {
        foo: { something: 42 },
        bar: { something: 42 },
      }
    });
    expect(s).toMatchSnapshot();
  });

  it("unifies identical objects", () => {
    const obj = { something: 42 };
    const s = stringify({
      foo: {
        foo: obj,
        bar: obj,
      }
    });
    expect(s).toMatchSnapshot();
  });

  it("unifies identical toplevel objects", () => {
    const obj = {
      version: "0.1.2",
    };
    const s = stringify({
      "foo@^0.1.0": obj,
      "foo@^0.1.1": obj,
      "foo@^0.2.0": {
        version: "0.2.0",
      },
    });
    expect(s).toMatchSnapshot();
  });
});

describe("wrapKey", () => {
  it("wraps key conditionally", () => {
    const src: (string | number | boolean)[] = [
      1234,
      "1234",
      "foo",
      "@babel/core",
      "webpack@^4.46.0",
      "webpack@4.x || 5.x",
      "https://example.com/",
      true,
      "true",
      "trues",
      "atrue",
      false,
      "false",
      "falsey",
      "afalse",
      "a\nb",
      "foo[bar]",
      "a\\b",
    ];
    const wrapped = src.map((k) => [k, wrapKey(k)] as const);
    expect(wrapped).toEqual([
      [1234, "1234"],
      ["1234", '"1234"'],
      ["foo", "foo"],
      ["@babel/core", '"@babel/core"'],
      ["webpack@^4.46.0", "webpack@^4.46.0"],
      ["webpack@4.x || 5.x", '"webpack@4.x || 5.x"'],
      ["https://example.com/", '"https://example.com/"'],
      [true, "true"],
      ["true", '"true"'],
      ["trues", '"trues"'],
      ["atrue", "atrue"],
      [false, "false"],
      ["false", '"false"'],
      ["falsey", '"falsey"'],
      ["afalse", "afalse"],
      ["a\nb", '"a\\nb"'],
      ["foo[bar]", '"foo[bar]"'],
      ["a\\b", '"a\\\\b"'],
    ]);
  });
});

describe("cmpAlpha", () => {
  it("Sorts strings alphabetically", () => {
    const sorted = sortWithPriority([
      "foo",
      "name",
      "ばー",
      "valueOf",
      "uid",
      "registry",
      "баз",
      "version",
      "dependencies",
      "hasOwnProperty",
      "resolved",
      "isPrototypeOf",
      "integrity",
    ]);
    expect(sorted).toEqual([
      "name",
      "version",
      "uid",
      "resolved",
      "integrity",
      "registry",
      "dependencies",
      "foo",
      "hasOwnProperty",
      "isPrototypeOf",
      "valueOf",
      "баз",
      "ばー",
    ]);
  });
});
