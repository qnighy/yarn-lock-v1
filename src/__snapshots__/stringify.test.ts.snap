// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify doesn't unify equivalent objects 1`] = `
"foo:
  bar:
    something 42
  foo:
    something 42
"
`;

exports[`stringify doesn't unify same primitives 1`] = `
"foo:
  bar 42
  foo 42
"
`;

exports[`stringify stringifies an empty object 1`] = `""`;

exports[`stringify stringifies an object with a single key 1`] = `
"foo:
"
`;

exports[`stringify stringifies package info 1`] = `
"call-bind@^1.0.2:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/call-bind/-/call-bind-1.0.2.tgz#b1d4e89e688119c3c9a903ad30abb2f6a919be3c\\"
  integrity sha512-7O+FbCihrB5WGbFYesctwmTKae6rOiIzmz1icreWJ+0aA7LJfuqhEso2T9ncpcFtzMQtzXf2QGGueWJGTYsqrA==
  dependencies:
    function-bind \\"^1.1.1\\"
    get-intrinsic \\"^1.0.2\\"

has-symbols@^1.0.1:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/has-symbols/-/has-symbols-1.0.2.tgz#165d3070c00309752a1236a479331e3ac56f1423\\"
  integrity sha512-chXa79rL/UC2KlX17jo3vRGz0azaWEx5tGqZg5pO3NUyEJVB17dMruQlzCCOfUvElghKcm5194+BCRvi2Rv/Gw==

is-regex@^1.1.2:
  version \\"1.1.2\\"
  resolved \\"https://registry.yarnpkg.com/is-regex/-/is-regex-1.1.2.tgz#81c8ebde4db142f2cf1c53fc86d6a45788266251\\"
  integrity sha512-axvdhb5pdhEVThqJzYXwMlVuZwC+FF2DpcOhTS+y/8jVq4trxyPgfcwIxIKiyeuLlSQYKkmUaPQJ8ZE4yNKXDg==
  dependencies:
    call-bind \\"^1.0.2\\"
    has-symbols \\"^1.0.1\\"
"
`;

exports[`stringify stringifies simple values 1`] = `
"foo:
  boolean false
  numeric 42
  s baz
  s2 \\"baz baz\\"
  sboolean \\"true\\"
  snumeric \\"42\\"
"
`;

exports[`stringify unifies identical objects 1`] = `
"foo:
  bar, foo:
    something 42
"
`;
